Deleting --spec-file argument from TestHarness.

Elley-Folks opened the pull request:
Refs Issue-#19107. --spec-file has overlapping functionality with --test-root. Removed --spec-file and function notMySpecFile that used it from TestHarness.py.

GiudGiud commented:
you'll want to try rebasing to avoid merge commits

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1514235/) on db49987 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24307/site)

This comment will be updated on new commits.

Elley-Folks committed:
--spec-file renamed to --run-dir

Deleted --test-root from argparse.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -358,11 +358,11 @@ def findAndRunTests(self, find_only=False):
-            self._infiles = [os.path.basename(self.options.spec_file)]
+        if self.options.run_dir and os.path.isdir(self.options.run_dir):
+            search_dir = self.options.run_dir
+        elif self.options.run_dir and os.path.isfile(self.options.run_dir):
```
This is the combination that I believe needs to go. We don't want to accept a file in "--run_dir". You can put the error message right here!

Elley-Folks replied:
This was done, thank you for the feedback!

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -1021,8 +1021,7 @@ def parseCLArgs(self, argv):
         parser.add_argument('--check-input', action='store_true', dest='check_input', help='Run check_input (syntax) tests only')
         parser.add_argument('--no-check-input', action='store_true', dest='no_check_input', help='Do not run check_input (syntax) tests')
-        parser.add_argument('--spec-file', action='store', type=str, dest='spec_file', help='Supply a path to the tests spec file to run the tests found therein. Or supply a path to a directory in which the TestHarness will search for tests. You can further alter which tests spec files are found through the use of -i and --re')
-        parser.add_argument('-C', '--test-root', nargs=1, metavar='dir', type=str, dest='spec_file', help='Tell the TestHarness to search for test spec files at this location.')
```
Add this `-C` option to your new `--run-dir`option.

Elley-Folks replied:
This was done, thanks for the feedback!

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1520027/) on 0631e83 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24307/coverage/xfem)

</details>



This comment will be updated on new commits.

Elley-Folks committed:
--run-dir and -i improvements

Added some logic to eliminate overlapping functionality with -i and --run-dir. -i no longer accepts a path. --run-dir, does not accept filename at end of path.

Elley-Folks committed:
Update test_ArbitrarySpecFile.py

Fixed unit test so it does not test for old behavior.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -1104,8 +1119,8 @@ def checkAndUpdateCLArgs(self):
-        if opts.spec_file and not os.path.exists(opts.spec_file):
-            print('ERROR: --spec-file supplied but path does not exist')
+        if opts.run_dir and not os.path.exists(opts.run_dir):
+            print('ERROR: --run-dir does not accept a file name, please enter a directory instead')
```
This error message is misleading: `exists` just checks that the path is valid, it doesn't differentiate between a directory and a file. So if you pass a directory that doesn't exist, you'll still get this error.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -357,16 +357,32 @@ def findAndRunTests(self, find_only=False):
-        elif self.options.spec_file and os.path.isfile(self.options.spec_file):
-            search_dir = os.path.dirname(self.options.spec_file)
-            self._infiles = [os.path.basename(self.options.spec_file)]
+            if(os.path.dirname(self._infiles[0]) !=''):
```
Oops! This is a critical error here! Indenting in Python is significant. This `if` statement is nested underneath the one above, so you'd only be testing this logic when `-i` is supplied, which is logically incorrect.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -357,16 +357,32 @@ def findAndRunTests(self, find_only=False):
+                print('ERROR entered a path, -i can only accept a file name or directory name.')
+                sys.exit(1)
+
+        #used for --run-dir
```
You don't need comments like this. They are self evident from the line below. I know you are following the pattern already established. If anything, they should all be stripped unless they are explaining something. No need to state things that can be verified on simple inspection.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -357,16 +357,32 @@ def findAndRunTests(self, find_only=False):
+        #used only when --run-dir <path> is used
+        elif self.options.run_dir and os.path.isfile(self.options.run_dir):
+
+            if(os.path.basename(self.options.run_dir[0]) !=''):
```
This is redudant with `isfile` - they are logically equivalent. I don't believe you can come up with a case where you make it into the error statement.

permcody commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -22,10 +22,10 @@ class QueueManager(Scheduler):
 
     Using this one unmodified job, the spec file involved is noted, and instructs the derived
-    scheduler how to launch this one single spec file (using --spec-file), along with any
+    scheduler how to launch this one single spec file (using --run-dir), along with any
```
So I was unaware of where we were using --spec-file. Apparently it's used in the PBS launcher right here and we are relying on it to run a single file. Two confounding issues:
- We aren't testing PBS, which is admittedly very bad, but if we were this would hopefully fail, BUT...
- We have a logic error where we don't actually look at the argument for `--run-dir` without `-i` so we are missing 100% of the use case here and it wouldn't fail even if we ran it!

permcody reviewed (changes requested):
Well this was a good PR, but admittedly trickier than I had anticipated. `--spec-file` is used and looking at what the PBS launcher does, I'm not convinced that we do need this special capability. I'm going to update the original ticket with a new design.

GiudGiud commented:
No
